### PR TITLE
Ensure redirect middleware preserves query strings

### PR DESCRIPTION
**Pull request by MattSegal**
This PR intends to resolve an issue (https://github.com/wagtail/wagtail/issues/4414) where Wagtail's RedirectMiddleware does not preserve querystring parameters. I believe that querystrings should be preserved on redirect.

Tests and linting pass locally.

My only concern with this PR is that prior authors of the redirect middleware were aware of the behaviour that I'm trying to fix. 

Looking at this relevant past issue (https://github.com/wagtail/wagtail/issues/647), I believe that my change doesn't conflict with the intent of the established logic, which stripped the querystring parameters to ensure that we can match a redirect with no params. My change preserves that behaviour.

**Acceptance Criteria**

* [ ] Querystring parameters stripped from request before matching
* [ ] Querystring parameters passed through to the redirect target
* [ ] If there are querystring parameters already specified in the `redirect_link`, then they override any params passed through in the original request

Eg. established behavior

```
old_path        /redirectme
redirect_link   /target?foo=1

request         /redirectme/
response        /target/?foo=1
```

My change intends to preserve this behaviour, but also handle this case

```
old_path        /redirectme
redirect_link   /target

request         /redirectme/?foo=1
response        /target/?foo=1
```

There's also this tricky corner case, where I default to the respecting the `redirect_link` 

```
old_path        /redirectme
redirect_link   /target?foo=1

request         /redirectme/?foo=2
response        /target/?foo=1
```

### CODE REVIEW

File: wagtail/contrib/redirects/tests.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def test_temporary_redirect(self):
         # Check that we were redirected temporarily
         self.assertRedirects(response, '/redirectto', status_code=302, fetch_redirect_response=False)
 
-    def test_redirect_stripping_query_string(self):
+    def test_redirect_matches_query_string(self):
```
Comment (kaedroho) [inline wagtail/contrib/redirects/tests.py]: Often, sites are migrated from a platform where the page is identified by a query parameter. So a user might add the following redirect:

``/index.php?page_id=1`` to ``/about-us/``

But it seems the logic here will actually redirect to ``/about-us/?page_id=1``.

Is it worth adding logic in to handle this?

Also, ideally, ``/index.php?page_id=1&utm_source=foo`` should forward to ``/about-us/?utm_source=foo``

**Comment by kaedroho** [comment]
Thanks for contributing this! I can't remember if there's a reason why we strip the query string other than helping it match more URLs and cleaning out crufty URL parameters.

The implementation as it stands looks good to me, just an open question about whether we should strip out URL parameters that we were expecting in the original URL.

File: wagtail/contrib/redirects/tests.py (hunk 1 above)
Comment (MattSegal) [reply wagtail/contrib/redirects/tests.py]: how would we know which querystring parameters to keep and which ones to throw away?

**Comment by MattSegal** [comment]
Added a defensive check against a missing `QUERYSTRING` key in `request.META`, which happened to me in testing a little while back.

File: wagtail/contrib/redirects/tests.py (hunk 1 above)
Comment (kaedroho) [reply wagtail/contrib/redirects/tests.py]: I'm thinking that any query parameters that were specified in the "old path" field should be thrown away, unless also specified in the redirect field. Everything else should be passed through.

This logic might be a little confusing to users though...

Comment (jjanssen) [reply wagtail/contrib/redirects/tests.py]: Regarding the discussed options, I think there are two actual use-cases:
1. You have rounded up a migration to Wagtail, therefore `/index.php?page-id=5` should redirect to `/about-us/`.
2. You are currently changing URLs in your Wagtail site, therefore `/previous-search-page/?q=keyword` should redirect to `/new-search-page/?q=keyword`.

The edge-case would be `/index.php?page-id=5&utm_source=12` but supporting that would be quite a hassle and would be really confusing for the end-user to configure.

To make it feasible for both scenarios it could make sense to add a checkbox whether the editor wants to preserve querystrings while configuring the redirect, the same way it asks whether it should permanently redirect.  

Curious to see both your perspectives on this.

Comment (MattSegal) [reply wagtail/contrib/redirects/tests.py]: @jjanssen I think it'd be possible to handle the edge case you mentioned without too much additional craziness.

The rules in english would be:

* Specified source params are stripped if not present in the target
* Unspecified params are passed through

```
from        /index.php?page-id=5
to          /about-us/
visit       /index.php?page-id=5
redirect    /about-us/

from        /index.php?page-id=5
to          /about-us/
visit       /index.php?page-id=5&utm_source=12
redirect    /about-us/?utm_source=12

from        /previous-search-page/
to          /new-search-page/
visit       /previous-search-page/?q=keyword
redirect    /new-search-page/?q=keyword

from        /previous-search-page/?q=keyword
to          /new-search-page/?q=keyword
visit       /previous-search-page/?q=keyword
redirect    /new-search-page/?q=keyword
```

what do you think?

Comment (gasman) [reply wagtail/contrib/redirects/tests.py]: @MattSegal 👍 That sounds like the ideal behaviour to me.

Comment (jjanssen) [reply wagtail/contrib/redirects/tests.py]: Sounds like a great idea Matt!

Comment (MattSegal) [reply wagtail/contrib/redirects/tests.py]: Just pushed some new changes. Turns out I couldn't do everything I said that I would do. Most of it, but not everything. I think/hope my tests make the expected behavior pretty clear.

## Can do:

Pass through requested params

```
from        /previous-search-page
to          /new-search-page
visit       /previous-search-page?q=keyword
redirect    /new-search-page?q=keyword
```

Strip old params:

```
from        /index.php?page-id=5
to          /about-us
visit       /index.php?page-id=5
redirect    /about-us
```

Merge requested and specified params:

```
from        /index.php
to          /about-us?foo=1
visit       /index.php?utm_source=12
redirect    /about-us?utm_source=12&foo=1
```

Respect redirected params

```
from        /previous-search-page?q=keyword
to          /new-search-page?q=keyword
visit       /previous-search-page?q=keyword
redirect    /new-search-page?q=keyword
```

## Can't do:

If you visit a URL that doesn't match the old_path querystring exactly, then we won't be able to find a redirect. Even if we could there are potential conflicts where 2 redirects could match 1 request. As such, we can't easily do this:

```
from        /index.php?page-id=5
to          /about-us/
visit       /index.php?page-id=5&utm_source=12
redirect    /about-us/?utm_source=12
```

Instead we have to 404, or look for a redirect that doesn't have a querystring.

**Comment by MattSegal** [comment]
Just pushed some new changes. Turns out I couldn't do everything I said that I would do. Most of it, but not everything. I think/hope my tests make the expected behavior pretty clear.

## Can do:

Pass through requested params

```
from        /previous-search-page
to          /new-search-page
visit       /previous-search-page?q=keyword
redirect    /new-search-page?q=keyword
```

Strip old params:

```
from        /index.php?page-id=5
to          /about-us
visit       /index.php?page-id=5
redirect    /about-us
```

Merge requested and specified params:

```
from        /index.php
to          /about-us?foo=1
visit       /index.php?utm_source=12
redirect    /about-us?utm_source=12&foo=1
```

Respect redirected params

```
from        /previous-search-page?q=keyword
to          /new-search-page?q=keyword
visit       /previous-search-page?q=keyword
redirect    /new-search-page?q=keyword
```

## Can't do:

If you visit a URL that doesn't match the old_path querystring exactly, then we won't be able to find a redirect. Even if we could there are potential conflicts where 2 redirects could match 1 request. As such, we can't easily do this:

```
from        /index.php?page-id=5
to          /about-us/
visit       /index.php?page-id=5&utm_source=12
redirect    /about-us/?utm_source=12
```

Instead we have to 404, or look for a redirect that doesn't have a querystring.

**Comment by gasman** [comment]
Hi @MattSegal - just wondering what the reason is for closing this? This is still something we're keen to fix, and was on our backlog to be reviewed...

**Comment by MattSegal** [comment]
Hi @gasman I'm thought that this PR had fallen by the wayside, given this is on your to-do list I'll re-open.

**Comment by MattSegal** [comment]
Given merge conflicts with wagtail/contrib/redirects/middleware.py I don't want to keep this open

**Comment by steverowland** [comment]
Hmm there is a proposed solution and now it's closed? We would need this one. There is no alternative?
